pkg/controller/nodeconfig/controller.go

```diff
@@ -0,0 +1,141 @@
+		ctx:              ctx,
+		NodeName:         nodeName,
+		NodeConfigs:      nodecfg,
+		NodeConfigsCache: nodecfg.Cache(),
+		WaitGroup:        &WaitGroup,
+	}
+
+	monitorNnumbers := len(toMonitorServices)
+	WaitGroup.Add(monitorNnumbers)
+
+	monitorModules := make([]interface{}, 0, monitorNnumbers)
+	for _, serviceName := range toMonitorServices {
+		monitorModule := monitor.InitServiceMonitor(ctx, &WaitGroup, nodecfg, nodes, nodeName, serviceName)
+		monitorModules = append(monitorModules, monitorModule)
+	}
+	monitor.StartsAllMonitors(monitorModules)
+
+	ctl.NodeConfigs.OnChange(ctx, HandlerName, ctl.OnNodeConfigChange)
+	ctl.NodeConfigs.OnRemove(ctx, HandlerName, ctl.OnNodeConfigRemove)
+
+	return ctl, nil
+}
+
+func (c *Controller) OnNodeConfigChange(node string, nodecfg *nodeconfigv1.NodeConfig) (*nodeconfigv1.NodeConfig, error) {
+	if nodecfg == nil || nodecfg.Spec.NodeName != c.NodeName {
+		logrus.Infof("Skip this round (OnChange) with NodeConfigs :%+v", nodecfg)
+		return nil, nil
+	}
+
+	// NTP related handling
+	ntpConfigHandler := config.NewNTPConfigHandler(nodecfg.Spec.NTPConfig)
+	if nodecfg.Status.NTPConditions != nil && nodecfg.Spec.NTPConfig.ConfigModified {
+		nodecfgCpy := nodecfg.DeepCopy()
+		config.UpdateNTPConfigChange(&nodecfgCpy.Status)
+		if !reflect.DeepEqual(nodecfg, nodecfgCpy) {
+			return c.NodeConfigs.Update(nodecfgCpy)
+		}
+	}
+
+	if config.CheckConfigApplied(config.NTPName, nodecfg.Status) {
+		if err := ntpConfigHandler.DoNTPUpdate(); err != nil {
+			logrus.Errorf("Update NTP config fail. err: %v", err)
+			c.NodeConfigs.EnqueueAfter(nodecfg.Namespace, nodecfg.Name, enqueueJitter())
+			return nil, err
+		}
+		if err := ntpConfigHandler.RestartService(); err != nil {
+			logrus.Errorf("Restart systemd-timesyncd fail. err: %v", err)
+			c.NodeConfigs.EnqueueAfter(nodecfg.Namespace, nodecfg.Name, enqueueJitter())
+			return nil, err
+		}
+		ntpserver := nodecfg.Spec.NTPConfig.NTPServers
+		if err := ntpConfigHandler.UpdateNTPConfigPersistence(ntpserver); err != nil {
+			logrus.Errorf("Update NTP config to OEM fail. err: %v", err)
+			c.NodeConfigs.EnqueueAfter(nodecfg.Namespace, nodecfg.Name, enqueueJitter())
+			return nil, err
+		}
+
+		nodecfgCpy := nodecfg.DeepCopy()
+		config.UpdateNTPConfigApplied(&nodecfgCpy.Status)
+		nodecfgCpy.Spec.NTPConfig.ConfigModified = false
+		if !reflect.DeepEqual(nodecfg, nodecfgCpy) {
+			return c.NodeConfigs.Update(nodecfgCpy)
+		}
+	}
+
+	if nodecfg.Status.AllConfigStatus != ConfigApplied {
+		nodecfgCpy := nodecfg.DeepCopy()
+		nodecfgCpy.Status.AllConfigStatus = ConfigApplied
+		if !reflect.DeepEqual(nodecfg, nodecfgCpy) {
+			return c.NodeConfigs.Update(nodecfgCpy)
+		}
+	}
+
+	return nil, nil
+}
+
+func (c *Controller) OnNodeConfigRemove(node string, nodecfg *nodeconfigv1.NodeConfig) (*nodeconfigv1.NodeConfig, error) {
+	if nodecfg == nil || nodecfg.Spec.NodeName != c.NodeName {
+		logrus.Infof("Skip this round (OnRemove) with NodeConfigs :%+v", nodecfg)
```

Comment:
We should return an error here or requeue the object here.

From my experience with pcidevices controller which also runs as a daemonset, when the controller pod on node matching `nodecfg.Spec.NodeName` is unavailable, the other nodes will process the object and return nil. The nil return from OnRemove instructs wrangler to remove the finalizer and object is ignored without performing the rollback operation. 

https://github.com/harvester/pcidevices/blob/master/pkg/controller/pcideviceclaim/pcideviceclaim_controller.go#L92

Reply:
Nice tips. We need to wait for the right controller to handle it! Thanks for the explain!